### PR TITLE
feat: Add trace package to track chains, blocks, and txs

### PR DESCRIPTION
**Pull request by DavidNix**
# Description, Motivation, and Context

With black box testing, reporting a failure is the main feature. Figuring out why something failed (i.e. debugging) is another problem and often difficult for block boxes.

The source of truth in most chains are the blocks. This PR adds a package to enable tracing tests, chains, blocks, and their transactions. SQL allows you to filter and focus on different test invocations, chains, etc.

This aims to replace logging blocks with -log-level=debug. Logging was an ok first pass, but they are extremely hard to parse through. Duplicate blocks are logged, and multiple chain blocks are mixed in.

This is a *package only* and I'm happy to add invocation of the package to this PR or later. I wanted to get feedback before I went further down the rabbit hole. 

## SQLite

I chose SQLlite because I wanted to capture the relationship between the domain entities. The ideal may be a NoSQL JSON-native embedded database with semi-relational features, but I wasn't able to find (nor have experience) with such a product. 

## Known Limitations, Trade-offs, Tech Debt
* The MVP user interface is a sqlite console where we expect the user to write SQL queries. I envision buidling a terminal UI on top the sqlite db to allow for quick block inspection. 
* Since txs are JSON, I still have to explore how to appropriately query that. SQLite has a JSON extention but it may require CGO.

### CODE REVIEW

File: internal/trace/migrate.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// Migrate migrates db in an idempotent manner.
+// If an error is returned, it's acceptable to delete the database and start over.
+// The basic ERD is as follows:
+//  ┌────────────────────┐          ┌────────────────────┐         ┌────────────────────┐          ┌────────────────────┐
+//  │                    │          │                    │         │                    │          │                    │
+//  │                    │         ╱│                    │        ╱│                    │         ╱│                    │
+//  │     Test Case      │───────┼──│       Chain        │───────○─│       Block        │────────○─│         Tx         │
+//  │                    │         ╲│                    │        ╲│                    │         ╲│                    │
+//  │                    │          │                    │         │                    │          │                    │
+//  └────────────────────┘          └────────────────────┘         └────────────────────┘          └────────────────────┘
```
Comment (DavidNix) [inline internal/trace/migrate.go:14]: ERD

File: internal/trace/sql.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package trace
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	_ "modernc.org/sqlite"
```
Comment (DavidNix) [inline internal/trace/sql.go:8]: This sqlite package is a pure Go implementation (does not require CGO). I've used it before and it's quite good.

**Comment by DavidNix** [comment]
I'll fix the linter errors on Tuesday when I return to work.

**Comment by mark-rushakoff** [comment]
> Request and/or assist with getting modernc/sqlite to use 3.38 

Actually, they're already on 3.38.x; I was reading the "Changelog" section of the godoc which is apparently out of date.

[JSON functions and operators](https://www.sqlite.org/json1.html) should just work AFAICT (but I haven't tried it myself).

File: internal/trace/doc.go [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Package trace saves data for inspection later to aid in debugging ibctest failures.
+package trace
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply internal/trace/doc.go:2]: Yeah, I like `blockdb`. I also tried `indexer`, `db` but didn't like those either.

File: internal/trace/sql_test.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+func migratedDB() *sql.DB {
+	db := emptyDB()
+	if err := Migrate(db); err != nil {
+		panic(err)
+	}
+	return db
+}
+
+func TestConnectDB(t *testing.T) {
+	f, err := os.CreateTemp("", t.Name())
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply internal/trace/sql_test.go:29]: In this case, I prefer my implementation to your alternative. Given this is only for a test. 1. Sharing the same file name for tests can lead to test pollution 2. os.CreateTemp creates a new filename each time.

File: internal/trace/sql_test.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package trace
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply internal/trace/sql_test.go:1]: For unit tests, I almost never use the `_test` package name. 

> the package name also ends with _test. When the package name ends like this, the test code can only access exported identifiers.

I often use unexported "test hooks" to aid in unit tests. Granted, I don't have such hooks in this PR, but they can be useful for mocking things out such as time.

File: internal/trace/chain.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+// Txs are transactions expected to be marshalled JSON.
+type Txs [][]byte
+
+// TraceBlock tracks a block at height with its transactions.
+// This method is idempotent and can be safely called multiple times with the same arguments.
+func (chain *Chain) TraceBlock(ctx context.Context, height int, txs Txs) error {
+	// TODO(nix 05-27-2022): Presentation in the database layer is generally bad practice. However, the first pass
+	// of this feature requires the user to make raw sql against the database. Therefore, to ease readability
+	// we indent json here. If we have a presentation layer in the future, I suggest removing the json indent here
```
[the remark this answers is not in the document]
Comment (DavidNix) [reply internal/trace/chain.go:25]: This indentation is what `go fmt` wants.

**Comment by DavidNix** [comment]
@mark-rushakoff Ready for 👀 again

**Review by mark-rushakoff — APPROVED** [review]
🚢 

(My indentation comment was referencing the json.Indent call mentioned in the comment, not the indentation of the Go comment in the file.)

**Comment by DavidNix** [comment]
> (My indentation comment was referencing the json.Indent call mentioned in the comment, not the indentation of the Go comment in the file.)

Gotcha. Yes, felt weird to do that. But for now, a sqlite console will be the first UI. 😬
